### PR TITLE
feat: add documentation and test for registryUrlTemplate

### LINKED ISSUE #8818: Support registryUrlTemplate
_Originally posted by @56KBs in https://github.com/renovatebot/renovate/discussions/8589_

> Hi,
> 
> I'm looking to run the regex manager on a resource which is hosted in a private GitLab instance. Is it possible to configure the registryUrl without using a capture group?
> 
> Here is some example configuration:
> 
> ```json
>     "regexManagers": [
>       {
>         "fileMatch": ["(^|/)Jenkinsfile$"],
>         "matchStrings": [
>           "@Library\\(\"jenkinsLib@(?<currentValue>.*?)\"\\)_"
>         ],
>         "datasourceTemplate": "gitlab-tags",
>         "depNameTemplate": "group/repositoryName",
>         "versioningTemplate": "semver",
>       }
>     ],
> ```
> 
> My issue is that `group/repositoryName` is hosted on `https://gitlab.my-private-instance.com`, meaning I need to configure a custom registry URL.
> 
> The configuration suggests this is possible using the registryUrl capture group, however in my scenario, I don't have text to match on that details the registry this resource is hosted on. I understand I could expand my regex to include the additional match and add the registry URL within my Jenkinsfile, however I am attempting to get this to function on a number of repositories without me having to update all of the repositories to include this additional information (Which may change in future too).
> 
> Is there any way to do what I'm looking to do, or would this be an enhancement to renovate?
> 
> Edit: From looking at how the code functions, it appears this functionality could be added via using a `registryUrlTemplate` configuration field which translates to the `registryUrls` as a single length array (Similar to the capture group itself). This obviously requires code changes to complete, so I was hoping there may be a way to do it without code changes - Although I'd happily contribute this functionality.
>

### PR DESCRIPTION
**Pull request by fgreinacher**
## Changes:

Seems like the support for what's asked for in #8818 has already been implemented as part of #7792.

This PR just adds documentation and a test case for this feature.

## Context:

Resolves #8818

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added unit tests, or
- [ ] No new tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

**Review by viceice — DISMISSED** [review]

**Comment by viceice** [comment]
renamed, as there is currently no docs update 😉

File: lib/config/common.ts [hunk 1]
```diff
@@ -136,6 +136,7 @@ export interface CustomManager {
   datasourceTemplate?: string;
   lookupNameTemplate?: string;
   versioningTemplate?: string;
+  registryUrlTemplate?: string;
```
Comment (fgreinacher) [inline lib/config/common.ts:139]: Not sure whether this is important?

File: lib/manager/common.ts [hunk 2]
```diff
@@ -37,6 +37,7 @@ export interface CustomExtractConfig extends ExtractConfig {
   lookupNameTemplate?: string;
   datasourceTemplate?: string;
   versioningTemplate?: string;
+  registryUrlTemplate?: string;
```
Comment (fgreinacher) [inline lib/manager/common.ts:40]: Same here: Not sure whether this is important?

**Review by fgreinacher — COMMENTED** [review]
> renamed, as there is currently no docs update 😉

Re-renamed, as now there is 😛 (I switched to `feat` because I think this makes it more obvious that new functionality is available)

File: lib/manager/regex/readme.md [hunk 3]
```diff
@@ -24,7 +24,7 @@ Configuration-wise, it works like this:
 - You can optionally have a `versioning` capture group or a `versioningTemplate` config field. If neither are present, `semver` will be used as the default
 - You can optionally have an `extractVersion` capture group or an `extractVersionTemplate` config field
 - You can optionally have a `currentDigest` capture group.
-- You can optionally have a `registryUrl` capture group. If it's a valid URL, it will be converted to the `registryUrls` field as a single-length array.
+- You can optionally have a `registryUrl` capture group or a `registryUrlTemplate` config field. If it's a valid URL, it will be converted to the `registryUrls` field as a single-length array.
```
Comment (viceice) [inline lib/manager/regex/readme.md:27]: I think one can have both 🤔

**Comment by viceice** [comment]
> > renamed, as there is currently no docs update 😉
> 
> Re-renamed, as now there is 😛 (I switched to `feat` because I think this makes it more obvious that new functionality is available)

So next time make a draft pr, so we can see it's not ready 😉

**Comment by 56KBs** [comment]
Hi @fgreinacher, I have raised #8611 to provide support for registryUrlTemplate because I don't believe support does already exist.

You can define `registryUrlTemplate` and have the result appear in the deps object as `registryUrl`, however this doesn't function as the regexManager is expecting `registryUrls` as an array, rather than a single string value.

**Comment by rarkins** [comment]
The regex manager should massage `registryUrl` to `registryUrls` as we do here: https://github.com/renovatebot/renovate/blob/cda66e08c533c400e8a44ccd23af31cd7e7de6ef/lib/manager/regex/__snapshots__/index.spec.ts.snap#L133-L135

Can we confirm if that's the case or not for `registryUrlTemplate` and fix if not?

**Comment by fgreinacher** [comment]
> The regex manager should massage `registryUrl` to `registryUrls` as we do here:
> 
> https://github.com/renovatebot/renovate/blob/cda66e08c533c400e8a44ccd23af31cd7e7de6ef/lib/manager/regex/__snapshots__/index.spec.ts.snap#L133-L135
> 
> Can we confirm if that's the case or not for `registryUrlTemplate` and fix if not?

It did not, I fixed it: https://github.com/renovatebot/renovate/pull/8819/commits/fc229224354b90c16a8ff574b2a5b72207a92951

**Comment by fgreinacher** [comment]
> Hi @fgreinacher, I have raised #8611 to provide support for registryUrlTemplate because I don't believe support does already exist.
> 
> You can define `registryUrlTemplate` and have the result appear in the deps object as `registryUrl`, however this doesn't function as the regexManager is expecting `registryUrls` as an array, rather than a single string value.

Oh no, so why did the same work -at least in a similar way 😉

It's really a pity that GitHub does not make it easier to link Issues and Discussions.

File: lib/manager/regex/readme.md (hunk 3 above)
Comment (fgreinacher) [reply lib/manager/regex/readme.md:27]: At least at the moment it's either-or. What use-case do you have in mind?

Comment (viceice) [reply lib/manager/regex/readme.md:27]: I know  you can use `versioning` capture group and `versioningTemplate` config at the same time, maybe the others too, so  i would expect it to work for `registryUrl` too.

```json
  {
  "versioningTemplate": "{{#if versioning}}{{versioning}}{{else}}semver{{/if}}",
  "registryUrlTemplate": "{{#if registryUrl}}{{registryUrl}}{{else}}https://x.y.test/path/any{{/if}}"
  }
```

**Comment by fgreinacher** [comment]
Looking deeper at https://github.com/renovatebot/renovate/pull/8611 I think that one is more mature than what I did, so I'm closing this MR.

Sorry for the confusion!

**Pull request closed without merging**